src_lockfree/skiplist.ml

```diff
@@ -0,0 +1,193 @@
+    if Random.bool () then cur_level
+    else if cur_level == sl.max_height then count_level 0
+    else count_level (cur_level + 1)
+  in
+  if sl.max_height = 0 then 0 else count_level 0
+
+let create ?(max_height = 10) () =
+  let max_height = Int.max max_height 1 in
+  let tail = create_new_node max (max_height - 1) in
+  let next =
+    Array.init max_height (fun _ -> Atomic.make { node = tail; marked = false })
+  in
+  let head = { key = min; height = max_height - 1; next } in
+  { head; max_height = max_height - 1 }
+
+(** Compares old_node and old_mark with the atomic reference and if they are the same then
+    Replaces the value in the atomic with node and mark *)
+let compare_and_set_mark_ref (atomic, old_node, old_mark, node, mark) =
+  let ({ node = current_node; marked = current_marked } as current) =
+    Atomic.get atomic
+  in
+  current_node == old_node && current_marked = old_mark
+  && ((current_node == node && current_marked = mark)
+     || Atomic.compare_and_set atomic current { node; marked = mark })
+
+(** Returns true if key is found within the skiplist else false;
+    Irrespective of return value, fills the preds and succs array with
+    the predecessors nodes with smaller key and successors nodes with greater than
+    or equal to key
+  *)
+let find_in (key, preds, succs, sl) =
+  let head = sl.head in
+  let rec iterate (prev, curr, succ, mark, level) =
+    if mark then
+      let snip =
+        compare_and_set_mark_ref (prev.next.(level), curr, false, succ, false)
+      in
+      if not snip then raise Failed_snip
+      else
+        let { node = new_succ; marked = mark } = Atomic.get succ.next.(level) in
+        iterate (prev, succ, new_succ, mark, level)
+    else if curr.key != max && curr.key < key then
+      let { node = new_succ; marked = mark } = Atomic.get succ.next.(level) in
+      iterate (curr, succ, new_succ, mark, level)
+    else (prev, curr)
+  in
+  let rec update_arrays prev level =
+    let { node = curr; marked = _ } = Atomic.get prev.next.(level) in
+    let { node = succ; marked = mark } = Atomic.get curr.next.(level) in
+    try
+      let prev, curr = iterate (prev, curr, succ, mark, level) in
+      preds.(level) <- prev;
+      succs.(level) <- curr;
+      if level > 0 then update_arrays prev (level - 1)
+      else if curr.key == max then false
+      else curr.key = key
+    with Failed_snip -> update_arrays head sl.max_height
+  in
+  update_arrays head sl.max_height
+
+(** Adds a new key to the skiplist sl. *)
+let add sl key =
+  let top_level = get_random_level sl in
+  let preds = create_dummy_node_array sl in
+  let succs = create_dummy_node_array sl in
+  let rec repeat () =
+    let found = find_in (key, preds, succs, sl) in
+    if found then false
+    else
+      let new_node_next =
+        Array.map
+          (fun element ->
+            let mark_ref = { node = element; marked = false } in
+            Atomic.make mark_ref)
+          succs
```

Comment:
Hmm... Aside from the array created here being too large, I believe there might be a more subtle space leak issue.

What happens here is that a new node is being constructed and as a part of that an array of references to other nodes is created.  A particular reference of this array is not updated during `add` nor are the references in the array created here subject to updates until a reference to the `new_node` is added to a predecessor node _at a specific level_.

Consider the following scenario:

1. A domain performing `add` is suspended after creating the `new_node_next` array.
2. Another domain `remove`s one of the nodes to which the `new_node_next` array has a reference at a non-zero level.
3. The domain performing `add` is resumed and completes the operation.

What will happen then that the `add` will notice that a successor node was removed at around line 118+ as the `compare_and_set_mark_ref` fails.  The `add` will then call `find_in` to update the `preds` and `succs`.  This will not, however, update the reference in the `new_node_next` array, which was created based on an earlier `succs`.  That is because the reference is at a level on which the `new_node` is not yet attached to the skip list (that is because the `compare_and_set_mark_ref` failed).

This means that after `add` returns, the `new_node` has been added to the skip list and the `new_node` contains a reference to a removed node.  This means that the `key` contained in that removed node cannot be garbage collected.  It will remain in memore until some call to `find_in` will notice that the removed node (due to marked references) and removes it.  But there is no guarantee such a call will happen. It might never happen.

Am I missing something?